keps/sig-api-machinery/4358-custom-resource-field-selectors/README.md

```diff
@@ -0,0 +1,1241 @@
+type SelectableField struct {
+	// jsonPath is a simple JSON path which is evaluated against each custom resource to produce a
+  // field selector value.
+  // Only JSON paths without the array notation are allowed.
+  // Must point to a field of type string, boolean or integer. Types with enum values
+  // and strings with formats are allowed.
+  // If jsonPath refers to absent field in a resource, the jsonPath evaluates to an empty string.
+  // Must not point to metdata fields.
+	JSONPath string `json:"jsonPath" protobuf:"bytes,1,opt,name=jsonPath"`
+}
+```
+
+Note that we do not offer a way to set a default value. We discussed this in API
+review and observed that because OpenAPI schema already provides a way to set
+a default value for a field, that we don't need (or want) to add a second way to set a
+default for selectable fields.  Instead, if the jsonPath refers to an absent field
+value in a resource, the selectable field value will be an empty string.
+
+##### Validation rules
+
+- `selectableFields[].jsonPath` must be a "simple path" (similar to
+  `additionalPrinterColumns[].jsonPath` except validated properly). I.e.
+    property names separated by '.', e.g. `.spec.x.y` (The path may not contain
+    map or array indexing)
+- The schema type referenced by `selectableFields[].jsonPath` must be one of:
+  string, integer, boolean (enums and fields with string formats are allowed).
+- There is a limit on the maximum `selectableFields` per CRD version that are allowed.
+- `selectableFields` may not refer to metadata fields.
+- `selectableFields` may not contain duplicates.
+
+
+#### OpenAPI Discovery
+
+We will add discovery details in to OpenAPIv3. For example:
+
+```json
+"org.example.v1.CustomResource": {
+  "type": "object",
+  "x-kubernetes-group-version-kind": [ ... ],
+  "x-kubernetes-selectable-fields": [
+    { "fieldPath": "metadata.name"},
+    { "fieldPath": "metadata.namespace"},
+    { "fieldPath": "spec.myField"},
+  ]
+},
+```
+
+#### Implementation
+
+Because `generic.StoreOptions` consumes selectors from built-in and CRDs in the
+same way, we would extend the `GetAttrs` of CRDs to offer fields for selection.
+https://github.com/jpbetz/kubernetes/tree/crd-object-filters contains a working
+prototype.
+
+We would leverage the fieldPath validation supported added for CRD Validation
+Rules to validate paths.
+
+We would leverage the jsonpath library in the same way as used by
+additionalPrinterColumns to retrieve values from custom resources using
+jsonpaths.
+
+The majority of the remaining work will be to update the API to add the
+field and to validate it.
+
+### User Stories
+
+- CustomResourceDefinition author wishes to provide filtered access by a field
+  - Author adds the field to an allow list in the CustomResourceDefinition
+  - If field is not always present, CustomResourceDefinition author declares how absent fields are handled
```

Comment:
If I see correctly in l.357-358 it's stated that an absent field will always evaluate to an empty string

(not sure if the user stories have to match what we want to implement)

Reply:
Ah, good catch. A prior version of the KEP handled this differently.  Fixed.